### PR TITLE
Add Python interface of prior_boxes

### PR DESCRIPTION
**Pull request by chengduoZH**
fix https://github.com/PaddlePaddle/Paddle/issues/8333
related issue https://github.com/PaddlePaddle/Paddle/issues/7488

### CODE REVIEW

File: python/paddle/v2/fluid/layers/nn.py [hunk 1]
```diff
@@ -64,7 +66,11 @@
     'nce',
     'beam_search',
     'row_conv',
+    'reshape',
```
Comment (qingqing01) [inline python/paddle/v2/fluid/layers/nn.py]: `reshape` has been exposed to Python in https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/layers/ops.py#L50

File: python/paddle/v2/fluid/layers/nn.py [hunk 2]
```diff
@@ -3091,3 +3097,311 @@ def multiplex(inputs, index):
...
+            'flip': flip,
+            'clip': clip,
+            'step_w': step_w,
+            'step_h': step_h,
+            'offset': offset
+        })
+    return box, var
+
+
+def prior_boxes(input_layers,
```
Comment (qingqing01) [inline python/paddle/v2/fluid/layers/nn.py]: I think only `prior_boxes` is enough. the definition of prior_box can be moved this function.

File: python/paddle/v2/fluid/layers/nn.py (hunk 1 above)
Comment (chengduoZH) [reply python/paddle/v2/fluid/layers/nn.py]: Done

File: python/paddle/v2/fluid/layers/nn.py [hunk 3]
```diff
@@ -3091,3 +3096,295 @@ def multiplex(inputs, index):
                 'Ids': index},
         outputs={'Out': [out]})
     return out
+
+
+def reshape_with_axis(input, axis):
```
Comment (qingqing01) [inline python/paddle/v2/fluid/layers/nn.py]: 我觉得不需要封装这个接口给用户，一个reshape足够了，类似这样的接口用户可以自己封装，不需要Fluid官方提供。

File: python/paddle/v2/fluid/layers/nn.py [hunk 4]
```diff
@@ -3091,3 +3096,295 @@ def multiplex(inputs, index):
...
+        new_shape += [reduce(mul, input_shape[axis[i]:axis[i + 1]], 1)]
+    new_shape = [-1] + new_shape
+
+    helper = LayerHelper('reshape', **locals())
+    out = helper.create_tmp_variable(helper.input_dtype())
+    helper.append_op(
+        type='reshape',
+        inputs={'X': [input]},
+        outputs={'Out': [out]},
+        attrs={'shape': new_shape})
```
Comment (qingqing01) [inline python/paddle/v2/fluid/layers/nn.py]:
```python
out = layers.reshape(input, shape=new_shape)
```

File: python/paddle/v2/fluid/layers/nn.py [hunk 5]
```diff
@@ -3091,3 +3096,295 @@ def multiplex(inputs, index):
...
+    if num_layer > 2:
+        step = int(math.floor(((max_ratio - min_ratio)) / (num_layer - 2)))
+        for ratio in xrange(min_ratio, max_ratio + 1, step):
+            min_sizes.append(base_size * ratio / 100.)
+            max_sizes.append(base_size * (ratio + step) / 100.)
+        min_sizes = [base_size * .10] + min_sizes
+        max_sizes = [base_size * .20] + max_sizes
+
+    if step_h:
+        assert isinstance(step_h,list) and len(step_h) == num_layer, \
```
Comment (qingqing01) [inline python/paddle/v2/fluid/layers/nn.py]: `assert` -> `raise ValueError('xxx')`

File: python/paddle/v2/fluid/layers/nn.py (hunk 4 above)
Comment (chengduoZH) [reply python/paddle/v2/fluid/layers/nn.py]: Done

File: python/paddle/v2/fluid/layers/nn.py [hunk 6]
```diff
@@ -3091,3 +3097,311 @@ def multiplex(inputs, index):
...
+
+          x = fluid.layers.data(name="data", shape=[3, 32, 32], dtype="float32")
+          reshaped = fluid.layers.reshape_with_axis(input=x, axis=2)
+          reshaped.shape
+            >> [-1, 1024]
+          reshaped = fluid.layers.reshape_with_axis(input=x, axis=[1,3])
+          reshaped.shape
+            >> [-1, 96, 32]
+    """
+    assert isinstance(axis, list), "axis should be list."
```
[the remark this answers is not in the document]
Comment (chengduoZH) [reply python/paddle/v2/fluid/layers/nn.py]: I have removed this function.

File: python/paddle/v2/fluid/layers/nn.py [hunk 7]
```diff
@@ -3091,3 +3094,41 @@ def multiplex(inputs, index):
...
+    Examples:
+       .. code-block:: python
+
+          x = fluid.layers.data(name="data", shape=[3, 32, 32], dtype="float32")
+          reshaped = fluid.layers.reshape_with_axis(input=x, axis=2)
+          reshaped.shape
+            >> [-1, 1024]
+    """
+    assert len(input.shape) > axis and axis > 0, \
+        "the axis should be litter than input.shape's."
```
Comment (pkuyym) [inline python/paddle/v2/fluid/layers/nn.py]: litter --> smaller
input.shape's --> arity of input's shape

File: python/paddle/v2/fluid/layers/nn.py [hunk 8]
```diff
@@ -3091,3 +3094,41 @@ def multiplex(inputs, index):
...
+
+def flatten(input, axis=1):
+    """
+    **Flatten Layer**
+
+    Flatten is used to reshape the input. After flattening, input's shape will
+    be two-dimension. The axis is used to indicate where to split the dimension.
+
+    Args:
+       input(variable): The input tensor.
```
Comment (pkuyym) [inline python/paddle/v2/fluid/layers/nn.py]: input(variable) --> input (Variable)

File: python/paddle/v2/fluid/layers/detection.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+                step_h=None,
+                offset=0.5,
+                variance=[0.1, 0.1, 0.1, 0.1],
+                flip=False,
+                clip=False,
+                name=None):
+    """
+    **Prior_boxes**
+
+    Generate prior boxes for SSD(Single Shot MultiBox Detector) algorithm.
```
Comment (pkuyym) [inline python/paddle/v2/fluid/layers/detection.py]: This comment is a little obscure, please refine it.

File: python/paddle/v2/fluid/layers/detection.py [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+    Each position of the inputs produces many prior boxes respectly, the number
+    of prior boxes which is produced by inputs respectly is determined by
+    the count of min_ratio, max_ratio and aspect_ratios, The size of the
+    box is in range(min_ratio, max_ratio) interval, which is generated in
+    sequence according to the aspect_ratios.
+
+    Args:
+       inputs(list): The list of input variables, the format of all variables is NCHW.
+       image(variable): The input image data of PriorBoxOp, the layout is NCHW.
```
Comment (pkuyym) [inline python/paddle/v2/fluid/layers/detection.py]: image(variable) --> image (Variable)

File: python/paddle/v2/fluid/layers/nn.py (hunk 8 above)
Comment (chengduoZH) [reply python/paddle/v2/fluid/layers/nn.py]: Done

File: python/paddle/v2/fluid/layers/detection.py (hunk 10 above)
Comment (chengduoZH) [reply python/paddle/v2/fluid/layers/detection.py]: Done

File: python/paddle/v2/fluid/layers/nn.py (hunk 7 above)
Comment (chengduoZH) [reply python/paddle/v2/fluid/layers/nn.py]: Done

File: python/paddle/v2/fluid/layers/detection.py (hunk 9 above)
Comment (chengduoZH) [reply python/paddle/v2/fluid/layers/detection.py]: Done

File: python/paddle/v2/fluid/layers/detection.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+from ops import reshape
+import math
+
+__all__ = [
+    'prior_box',
+    'prior_boxes',
+]
+
+
+def prior_boxes(inputs,
```
Comment (qingqing01) [inline python/paddle/v2/fluid/layers/detection.py]: prior_boxes -> prior_box

File: python/paddle/v2/fluid/layers/detection.py [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+
+from ..layer_helper import LayerHelper
+from ..framework import Variable
+from tensor import concat
+from ops import reshape
+import math
+
+__all__ = [
+    'prior_box',
+    'prior_boxes',
```
Comment (qingqing01) [inline python/paddle/v2/fluid/layers/detection.py]: remove `prior_boxes`

File: python/paddle/v2/fluid/tests/object_detection/test_prior_boxes.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        assert var_arr.shape[1] == 4
+        assert box_arr.shape[0] == var_arr.shape[0]
+
+
+class TestFitALine(unittest.TestCase):
+    def test_cpu(self):
+        main(use_cuda=False)
+
+    def test_cuda(self):
+        main(use_cuda=True)
```
Comment (qingqing01) [inline python/paddle/v2/fluid/tests/object_detection/test_prior_boxes.py]: 1. Need to move test to https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/tests/test_detection.py

2. box and var shape can be checked without exe.run

Comment (chengduoZH) [reply python/paddle/v2/fluid/tests/object_detection/test_prior_boxes.py]: Done

**Pull request merged**